lib/dex-oracle/driver.rb

```diff
@@ -47,38 +47,18 @@ def install(dex)
       raise "#{Resources.dx} does not exist and is required for DexMerger" unless File.exist?(Resources.dx)
       raise "#{Resources.driver_dex} does not exist" unless File.exist?(Resources.driver_dex)
       logger.debug("Merging input DEX (#{dex.path}) and driver DEX (#{Resources.driver_dex}) ...")
-      tf = Tempfile.new(%w(oracle-driver .dex))
-      cmd = "java -cp #{Resources.dx} com.android.dx.merge.DexMerger #{tf.path} #{dex.path} #{Resources.driver_dex}"
-      stdout, stderr = exec(cmd)
-      # Ideally, it says something like this:
-      # Merged dex #1 (36 defs/87.9KiB)
-      # Merged dex #2 (1225 defs/1092.7KiB)
-      # Result is 1261 defs/1375.0KiB. Took 0.2s
-      # But it might say this:
-      # Exception in thread "main" com.android.dex.DexIndexOverflowException: Cannot merge new index 65776 into a non-jumbo instruction!
-      if stderr.start_with?('Exception in thread "main"')
-        logger.error("Failure to merge input DEX and driver DEX:\n#{stderr}")
-        if stderr.include?('DexIndexOverflowException')
-          logger.error("Your input DEX inexplicably contains const-string and const-string/jumbo. This probably means someone fucked with it. In any case, it means DexMerge is failing because there are too many strings.\nTry this: baksmali the DEX, replace all const-string instructions with const-string/jumbo, then recompile with smali and use that DEX as input. Sorry, I don't want to do this for you. It's too complicated.")
-        end
-        exit -1
-      end
-      tf.close
-
       # Zip merged dex and push to device
       logger.debug('Pushing merged driver to device ...')
       tz = Tempfile.new(%w(oracle-driver .zip))
       # Could pass tz to create_zip, but Windows doesn't let you rename if file open
       # And zip internally renames files when creating them
       tempzip_path = tz.path
       tz.close
-      Utility.create_zip(tempzip_path, 'classes.dex' => tf)
+      Utility.create_zip(tempzip_path, 'classes.dex' => Resources.driver_dex, 'classes2.dex' => dex)
```

Comment:
WHYTHISWORKS: This adds the dex file of the apk and the dex file of the driver into a zip file which then gets uploaded to the device. The driver loads all classes from that zip file so it also finds the ones from the apk.

Reply:
Does _Android_ load classes2.dex automatically? I can't test it myself right now and I don't see the code to load it.

Reply:
I think app_process does that because as far as I know normal apks can also contain multiple dex files which are named the same way. But you should check that before merging.

Reply:
Please check + verify my implementation. I had to use an android project as the java project could not use DexClassLoader.